src/main/java/com/google/sps/servlets/StudyServlet.java

```diff
@@ -0,0 +1,200 @@
+import java.util.Comparator;
+import org.joda.time.DateTimeUtils;
+
+/**
+ * Handles the study mode questions and responses by adjusting and sorting a card's based on their
+ * familiarity score and last time seen.
+ */
+@WebServlet("/study")
+public class StudyServlet extends HttpServlet {
+
+  public class QuizCard {
+    private String quizWord;
+    private String cardKey;
+    private ArrayList<String> possibleResponses = new ArrayList<>();
+    private String correctAnswer;
+
+    public QuizCard(Card card) {
+      this.quizWord = card.getTextTranslated();
+      this.cardKey = card.getCardKey();
+      this.correctAnswer = card.getRawText().toLowerCase();
+      this.possibleResponses = WordSearch.generateWordOptions(this.correctAnswer);
+    }
+  }
+
+  private int numOfCardsPerRound = 1;
+  private int maxNumOfRounds = 4;
+
+  // Returns QuizCard's that are sorted by familiarity score as json
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    UserService userService = UserServiceFactory.getUserService();
+    List<Card> userCards = new ArrayList<>();
+    // Handles a logged out user
+    if (!userService.isUserLoggedIn()) {
+      response.sendRedirect("/StudyMode");
+      return;
+    }
+
+    String folderKey = request.getParameter("folderKey");
+    Query cardQuery = new Query("Card").setAncestor(KeyFactory.stringToKey(folderKey));
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    PreparedQuery results = datastore.prepare(cardQuery);
+
+    if (results != null) {
+      for (Entity entity : results.asIterable()) {
+        // If the card has no familiarity score then it's set to the default value
+        Entity updatedEntity = initializeCard(entity);
+        Card storedCard = storeCard(updatedEntity);
+        userCards.add(storedCard);
+      }
+    }
+    // Sorts cards in increasing order of familiarity score
+    userCards.sort(
+        (Card c1, Card c2) -> Double.compare(c1.getFamiliarityScore(), c2.getFamiliarityScore()));
+    List<List<QuizCard>> quiz = createQuizRounds(userCards);
+
+    Gson gson = new Gson();
+    String jsonResponse = gson.toJson(quiz);
+    response.setContentType("application/json;");
+    response.getWriter().println(jsonResponse);
+  }
+
+  // Updates and stores a card's new familiarity score
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    Entity cardEntity;
+    Double newScore;
+    Boolean answeredCorrectly = Boolean.parseBoolean(request.getParameter("answeredCorrectly"));
+    String cardKey = request.getParameter("cardKey");
+    long time = DateTimeUtils.currentTimeMillis();
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    try {
+      cardEntity = datastore.get(KeyFactory.stringToKey(cardKey));
+    } catch (EntityNotFoundException e) {
+      cardEntity = null;
```

Comment:
can't you return here instead of setting to null and checking below?

Reply:
done.